### ISSUE TITLE
Multiple person cards selected if card is dragged quickly

### ISSUE THREAD

**Issue by m133225** [HubTurbo/addressbook#291]
Steps to reproduce:
- Click on the bottom-most card
- Quickly drag the top-most card down at least by 1 card  
Result: 2 cards are selected, `IndexOutOfBoundsException` is thrown

This should work with any card, just indicated top/bottom-most for simplicity.

**Comment by m133225** [comment]
@yl-coder Might also want to address this issue in #287

**Comment by yl-coder** [comment]
Can't reproduce this on my computer on master branch. However when i was implementing the multiple selection, this occurs and I did something on #287. Can you try PR #287 on your computer to see if you can reproduce this?

**Comment by m133225** [comment]
Seems to be resolved in #287 (no practical issues, dragging works fine, no multiple selection). However, the exception is still being thrown.

**Comment by yl-coder** [comment]
No issue on my com. Could you post the stacktrace here?

**Comment by m133225** [comment]
```
Exception in thread "JavaFX Application Thread" java.lang.IndexOutOfBoundsException
	at com.sun.javafx.scene.control.ReadOnlyUnbackedObservableList.subList(ReadOnlyUnbackedObservableList.java:136)
	at javafx.collections.ListChangeListener$Change.getAddedSubList(ListChangeListener.java:242)
	at com.sun.javafx.scene.control.behavior.ListViewBehavior.lambda$new$177(ListViewBehavior.java:269)
	at javafx.collections.WeakListChangeListener.onChanged(WeakListChangeListener.java:88)
	at com.sun.javafx.collections.ListListenerHelper$Generic.fireValueChangedEvent(ListListenerHelper.java:329)
	at com.sun.javafx.collections.ListListenerHelper.fireValueChangedEvent(ListListenerHelper.java:73)
	at com.sun.javafx.scene.control.ReadOnlyUnbackedObservableList.callObservers(ReadOnlyUnbackedObservableList.java:75)
	at javafx.scene.control.MultipleSelectionModelBase.clearAndSelect(MultipleSelectionModelBase.java:378)
	at javafx.scene.control.ListView$ListViewBitSetSelectionModel.clearAndSelect(ListView.java:1403)
	at com.sun.javafx.scene.control.behavior.CellBehaviorBase.simpleSelect(CellBehaviorBase.java:256)
	at com.sun.javafx.scene.control.behavior.CellBehaviorBase.doSelect(CellBehaviorBase.java:240)
	at com.sun.javafx.scene.control.behavior.CellBehaviorBase.mousePressed(CellBehaviorBase.java:150)
	at com.sun.javafx.scene.control.skin.BehaviorSkinBase$1.handle(BehaviorSkinBase.java:95)
	at com.sun.javafx.scene.control.skin.BehaviorSkinBase$1.handle(BehaviorSkinBase.java:89)
	at com.sun.javafx.event.CompositeEventHandler$NormalEventHandlerRecord.handleBubblingEvent(CompositeEventHandler.java:218)
	at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:80)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
	at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
	at javafx.event.Event.fireEvent(Event.java:198)
	at javafx.scene.Scene$MouseHandler.process(Scene.java:3757)
	at javafx.scene.Scene$MouseHandler.access$1500(Scene.java:3485)
	at javafx.scene.Scene.impl_processMouseEvent(Scene.java:1762)
	at javafx.scene.Scene$ScenePeerListener.mouseEvent(Scene.java:2494)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:380)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:294)
	at java.security.AccessController.doPrivileged(Native Method)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleMouseEvent$354(GlassViewEventHandler.java:416)
	at com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:389)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.handleMouseEvent(GlassViewEventHandler.java:415)
	at com.sun.glass.ui.View.handleMouseEvent(View.java:555)
	at com.sun.glass.ui.View.notifyMouse(View.java:937)
```

**Comment by HansNewbie** [comment]
Maybe Mac only problem. @yl-coder remember when we tried to drag card around fast in my Mac?

I can try on my machine. Which branch should I try this?

**Comment by yl-coder** [comment]
Yea, maybe my com can't reproduce this. #287 [branch](https://github.com/HubTurbo/addressbook/tree/286-multiple-selection_future-tags-delete)

**Comment by m133225** [comment]
Either branch will do.

Yes, I think this is a mac-only problem. I can only reproduce it by dragging the card using the 3-finger gesture for Mac which does not involve clicking. Could be an erroneous assumption that the user needs to click on a card to drag.

**Comment by HansNewbie** [comment]
I get `IndexOutOfBoundsException` as well. I can get the same exception with mouse, although very rarely (about 1 out of 30 tries or so).

What I noticed is that the last card was still on focus when this happens. The card is not coloured blue, but it has blue surroundings (which is why I assume focus) although we are dragging the top card. Will see if can get pictures.

**Comment by HansNewbie** [comment]
Managed to get higher chance on using mouse. It seems to be processing time bug. I managed to reproduce the bug sometimes by dragging and then clicking the dragged card right away. Even exact step may not create the bug.

I assume this is because the list backing the card view has not finished updating but we already click something, in which the list may not have it yet due to the dragging. It only happens when the processor is choked (say when GC is working) hence not always reproducible.

Scenario I have in mind:
3 card, drag 1st card to between 2nd card and 3rd card. After dropping the card, the list now is of length 4 (temporarily) and the last card is actually index 4. Then upon removing the original 1st card, the list now is back to length 3, hence index 4 is out of bounds.

**Comment by yl-coder** [comment]
I have updated  #287 . to clear the selectionmodel before moving the element. Can help to see if still happen?

**Comment by m133225** [comment]
Still happens.

**Comment by yl-coder** [comment]
Hmm. I guess I should change the implementation of the ordered list. Not to remove and insert the list, but to keep an index of mapping.

**Comment by yl-coder** [comment]
@m133225 If i change it at my side, is it easy to port it over to ur side? Or should i wait till you merged ur PR?

**Comment by m133225** [comment]
Just go ahead and change it. I think my PR will take awhile anyway.

**Comment by HansNewbie** [comment]
I don't see it in the latest #287 with mousepad gesture with 3 cards.

**Comment by m133225** [comment]
@HansNewbie No Exception thrown?

**Comment by HansNewbie** [comment]
No.

**Comment by yl-coder** [comment]
I have changed something. could you guys try again

**Comment by HansNewbie** [comment]
If you are talking about c8022cadb53f43b4e54c2c5151b14a6012768e91, yes I have tried it and no exception thrown.

**Comment by yl-coder** [comment]
Yes. Ok case closed :p

**Comment by yl-coder** [comment]
@m133225 any exception thrown at ur side?

**Comment by m133225** [comment]
It still does throw occasionally, but seems like could be a different cause. Can't really reproduce it reliably.

**Comment by HansNewbie** [comment]
Was this corrected in #298?

**Comment by yl-coder** [comment]
Corrected in PR #330
